File: solutions/lua/robot-name/1/robot-name.lua
local all = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
local function _reset(l) return all:sub(l, l) .. all:sub(l, l) .. tostring(math.random(100, 999)) end
return {
  new = function()
    return {
      name = _reset(math.random(#all)),
      reset = function(self) self.name = _reset(math.random(#all)) end,
    }
  end
}

File: solutions/clojure/robot-name/1/src/robot_name.clj
(ns robot-name)

(def ^:private all-chars "ABCDEFGHIJKLMNOPQRSTUVWXYZ")
(def ^:private all-used (atom #{}))
(defn- gen-name []
  (let [name (str (rand-nth all-chars)
                  (rand-nth all-chars)
                  (format "%03d" (rand-int 1000)))
        [prev-used] (swap-vals! all-used conj name)]
    (if (prev-used name) (recur) name)))
(defn robot [] (atom (gen-name)))
(defn robot-name [robot] @robot)
(defn reset-name [robot] (reset! robot (gen-name)))
